### PR TITLE
[MANOPD-89259] Use_ssh_password_instead_of_ssh_key_in_k8s_installation

### PR DESCRIPTION
**Pull request by nikhil1697**
### Description
 This change aims to enhance the kubemarine system by introducing support for both SSH key and SSH password authentication methods during installation, maintenance, and checks. Currently, only SSH key authentication is supported, and this change will provide users with the flexibility to choose their preferred authentication method.
* 

Fixes # (issue)
The current kubemarine system only supports SSH key authentication, which restricts users who prefer to use passwords for authentication or have systems that do not support SSH key-based authentication.

### Solution
Updated Authentication Logic: Modify the code to handle both SSH key and SSH password authentication methods based on user-provided connection details.
* 

### Test Cases

**TestCase 1**

1. Provide password instead of keyfile in cluster.yaml and proceed with kubemarine tasks.

```
node_defaults:
  username: "centos"
  password: '{{ env.PASS }}' 
```
  Run "kubemarine check_iaas"
Results:

**Before** - 
`kubemarine check_iaas
Loading inventory file 'cluster.yaml'
2023-07-06 18:22:09,484 INFO Running action 'check iaas'
Excluded tasks:
        No excluded tasks
2023-07-06 18:22:10,159 ERROR Inventory file is failed to be validated against the schema.
2023-07-06 18:22:10,201 CRITICAL FAILURE!
2023-07-06 18:22:10,201 CRITICAL Property name 'password' is not one of ['keyfile', 'username', 'connection_port', 'connection_timeout', 'labels', 'taints', 'gateway']
2023-07-06 18:22:10,201 CRITICAL On inventory['node_defaults']`

**After** 
Check_iaas runs successfully

**TestCase 2**

Provide both keyfile and password in cluster.yaml and run Kubemarine tasks.

```
node_defaults:
  username: "centos"
  keyfile: "~/.ssh/id_rsa"
  password: '{{ env.PASS }}' 
```
  Run "kubemarine install"
  
Results:
**Before** - 
  ` kubemarine install
Loading inventory file 'cluster.yaml'
2023-07-06 18:42:06,303 INFO Running action 'install'
Excluded tasks:
        No excluded tasks
2023-07-06 18:42:06,620 ERROR Inventory file is failed to be validated against the schema.
2023-07-06 18:42:06,621 CRITICAL FAILURE!
2023-07-06 18:42:06,622 CRITICAL Property name 'password' is not one of ['keyfile', 'username', 'connection_port', 'connection_timeout', 'labels', 'taints', 'gateway']
2023-07-06 18:42:06,622 CRITICAL On inventory['node_defaults']`
  
**After** 
k8s installation runs successfully

### Checklist
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Integration CI passed
- [ ] Unit tests. If Yes list of new/changed tests with brief description
- [ ] There is no merge conflicts

### CODE REVIEW

File: documentation/Installation.md [hunk 1]
```diff
@@ -577,21 +579,23 @@ node:
...
 ```
 
 Following are the parameters allowed to be specified in the `node_defaults` section:
-* keyfile, username, connection_port, connection_timeout and gateway.
+* keyfile, password, username, connection_port, connection_timeout and gateway.
 * labels, and taints - specify at global level only if the [Mini-HA Scheme](#mini-ha-scheme) is used.
-
 For more information about the listed parameters, refer to the following section.
 
+Note: To establish an ssh connection, you can use either a keyfile or a password.
```
Comment (Imadzuma) [inline documentation/Installation.md:598]: Please specify more info, about `keyfile` and `password` fields: different configurations, what happens, if user specifies both of them, the limitations, that @ilia1243 discussed, and so on.

File: documentation/Installation.md [hunk 2]
```diff
@@ -1477,11 +1488,14 @@ By default, the installer uses the following parameters:
 |podPidsLimit|4096|
 |maxPods|110|
 |cgroupDriver|systemd|
+|serializeImagePulls|false|
```
Comment (Imadzuma) [inline documentation/Installation.md:1491]: Is it OK, that you specified that parameter in this PR? 
If yes, please provide info about it in PR description

File: kubemarine/core/connections.py [hunk 3]
```diff
@@ -56,14 +60,12 @@ def _create_connection_from_details(self, ip: str, conn_details: dict,
...
-                "key_filename": os.path.expanduser(conn_details['keyfile'])
-            },
+            connect_kwargs=creds,
             inline_ssh_env=inline_ssh_env
         )
 
     def _create_connection(self, ip: str, node: dict) -> fabric.connection.Connection:
-        if node.get('keyfile') is None:
+        if node.get('keyfile') is None and node.get('password') is None:
             raise Exception('There is no keyfile specified in configfile for node \'%s\'' % node['name'])
```
Comment (koryaga) [inline kubemarine/core/connections.py:69]: neither password

File: documentation/Installation.md [hunk 4]
```diff
@@ -577,21 +579,23 @@ node:
...
     username: "centos"
     keyfile: "/home/username/another.key"
+    password: "your_password"
     internal_address: "192.168.0.2"
     roles: ["control-plane"]
 ```
 
 Following are the parameters allowed to be specified in the `node_defaults` section:
-* keyfile, username, connection_port, connection_timeout and gateway.
+* keyfile, password, username, connection_port, connection_timeout and gateway.
```
Comment (koryaga) [inline documentation/Installation.md:594]: pls add a info and a reference to  ENV usage

File: kubemarine/core/connections.py [hunk 5]
```diff
@@ -44,9 +44,13 @@ def get_connection(self, ip: str) -> fabric.connection.Connection:
...
-        # fabric / invoke use default encoding of deployer.
-        # We need to use encoding of remote nodes to correctly encode output of remote commands.
-        # Currently remote nodes are always Linux, so hard-code 'utf-8'.
+
+        creds={}
+        if conn_details.get('keyfile'):
+            creds['key_filename'] = os.path.expanduser(conn_details['keyfile'])
+        elif conn_details.get('password'):
+            creds['password'] = conn_details.get('password')
+        #else:
```
Comment (Imadzuma) [inline kubemarine/core/connections.py:53]: If user specified `keyfile` in `node_defaults` and `password` in `nodes` for specific node, kubemarine takes `keyfile` value for ssh connection and ignores node's `password`.
If it's OK, please mention it in IN guide.

And remove this debug `else`

File: documentation/Installation.md (hunk 2 above)
Comment (nikhil1697) [reply documentation/Installation.md:1493]: Actually i didn't add this part this in this PR,  may be it came from Main when i pulled it and when i checked it was already there in main branch, so left it untoched.

**Review by koryaga — APPROVED** [review]

**Pull request merged**